### PR TITLE
Add support to parse gzipped EDIF files

### PR DESCRIPTION
**Pull request by clavin-xlnx**
This will allow the EDIF Parser (both serial and parallel) to accept gzipped EDIF files for parsing.  It adds this capability by checking the file extension, if it ends with `.gz`, it is assumed gzip compressed and will temporarily decompress it in the same directory as the original file and parse it.  Once parsing is completed, the temporary decompressed EDIF file is deleted.  This will overwrite an existing EDIF file if the provided `.edf.gz` file is given for parsing.

### CODE REVIEW

**Comment by clavin-xlnx** [comment]
@eddieh-xlnx - Should we support gzipped files that don't end in `.gz`? Currently, this implementation does not.

**Comment by clavin-xlnx** [comment]
Sorry @eddieh-xlnx , just found a performance bug, will mark as ready for review when fixed.

**Comment by clavin-xlnx** [comment]
Some performance numbers:

image

File: src/com/xilinx/rapidwright/edif/EDIFTools.java [hunk 1]
```diff
@@ -755,7 +755,7 @@ public static EDIFNetlist loadEDIFFile(Path fileName) {
     public static EDIFNetlist loadEDIFFile(Path fileName, int maxThreads) {
         try {
             final long size = Files.size(fileName);
-            if (ParallelEDIFParser.calcThreads(size, maxThreads) > 1) {
+            if (ParallelEDIFParser.calcThreads(size, maxThreads, fileName.toString().endsWith(".gz")) > 1) {
```
Comment (eddieh-xlnx) [inline src/com/xilinx/rapidwright/edif/EDIFTools.java:758]:
```suggestion
            long estimatedUncompressedSize = size * EDIF_GZIP_RATIO; 
            if (ParallelEDIFParser.calcThreads(estimatedUncompressedSize, maxThreads) > 1) {
```
(After moving the `EDIF_GZIP_RATIO` member over from `ParallelEDIFParser`)
Could this be more meaningful/generic?

File: src/com/xilinx/rapidwright/edif/ParallelEDIFParser.java [hunk 2]
```diff
@@ -105,7 +110,11 @@ protected void initializeWorkers() throws IOException {
     private int numberOfThreads;
 
     public EDIFNetlist parseEDIFNetlist() throws IOException {
-        return parseEDIFNetlist(CodePerfTracker.SILENT);
+        EDIFNetlist netlist = parseEDIFNetlist(CodePerfTracker.SILENT);
+        if (this.fileName.toString().endsWith(".gz")) {
```
Comment (eddieh-xlnx) [inline src/com/xilinx/rapidwright/edif/ParallelEDIFParser.java:114]:
```suggestion
        if (fileName.toString().endsWith(".gz")) {
```

File: src/com/xilinx/rapidwright/util/FileTools.java [hunk 3]
```diff
@@ -1832,5 +1833,62 @@ public static void main(String[] args) {
...
+     */
+    public static Path decompressGZIPFile(Path gzipFile) {
+        String fileNameStr = gzipFile.toString();
+        if (!fileNameStr.endsWith(".gz")) return null;
+        Path target = FileTools.getDecompressedGZIPFileName(gzipFile);
+        // Using a larger buffer size for GZIPInputStream improved runtime 5-10%
+        try (GZIPInputStream gis = new GZIPInputStream(new FileInputStream(fileNameStr), 65536)) {
+            Files.copy(gis, target, StandardCopyOption.REPLACE_EXISTING);
+        } catch (FileNotFoundException e) {
+            throw new UncheckedIOException(e);
```
Comment (eddieh-xlnx) [inline src/com/xilinx/rapidwright/util/FileTools.java:1855]:
```suggestion
```
`FileNotFoundException` is a subclass of `IOException`

File: src/com/xilinx/rapidwright/util/FileTools.java [hunk 4]
```diff
@@ -1832,5 +1833,62 @@ public static void main(String[] args) {
...
+     * Compresses the provided file using GZIP (adds '.gz' extension) 
+     * @param uncompressedFile The path to the uncompressed file
+     * @return The path to the compressed file which has a '.gz' extension
+     */
+    public static Path compressFileUsingGZIP(Path uncompressedFile) {
+        Path compressedFile = Paths.get(uncompressedFile.toString() + ".gz");
+        try (GZIPOutputStream gos = new GZIPOutputStream(new FileOutputStream(compressedFile.toFile()), 65536)) {
+            Files.copy(uncompressedFile, gos);
+        } catch (FileNotFoundException e) {
+            throw new UncheckedIOException(e);
```
Comment (eddieh-xlnx) [inline src/com/xilinx/rapidwright/util/FileTools.java:1872]:
```suggestion
```

File: test/src/com/xilinx/rapidwright/edif/TestEDIFParser.java [hunk 5]
```diff
@@ -119,4 +122,30 @@ public void loadEDIFSingleThreaded() throws IOException {
...
+
+    @Test
+    public void testGZIPEDIFParsing(@TempDir Path tempDir) throws IOException {
+        Path src = tempDir.resolve(input.getFileName());
+        Files.copy(input, src);
+        Path compressed = FileTools.compressFileUsingGZIP(src);
+        Assertions.assertTrue(compressed.toString().endsWith(".gz"));
+        Assertions.assertTrue(src.toFile().exists());
+        Files.delete(src);
+        EDIFTools.loadEDIFFile(compressed);
```
Comment (eddieh-xlnx) [inline test/src/com/xilinx/rapidwright/edif/TestEDIFParser.java:134]:
```suggestion
        Assertions.assertFalse(src.toFile().exists());
        Assertions.assertNotNull(EDIFTools.loadEDIFFile(compressed));
```

File: test/src/com/xilinx/rapidwright/edif/TestEDIFParser.java [hunk 6]
```diff
@@ -119,4 +122,30 @@ public void loadEDIFSingleThreaded() throws IOException {
...
+    public void testGZIPEDIFParsingParallel(@TempDir Path tempDir) throws IOException {
+        Path src = tempDir.resolve(input.getFileName());
+        Files.copy(input, src);
+        Path compressed = FileTools.compressFileUsingGZIP(src);
+        Assertions.assertTrue(compressed.toString().endsWith(".gz"));
+        Assertions.assertTrue(src.toFile().exists());
+        Files.delete(src);
+        try (ParallelEDIFParser p = new ParallelEDIFParser(compressed)) {
+            p.parseEDIFNetlist();
+        }
```
Comment (eddieh-xlnx) [inline test/src/com/xilinx/rapidwright/edif/TestEDIFParser.java:148]:
```suggestion
        Assertions.assertFalse(src.toFile().exists());
        try (ParallelEDIFParser p = new ParallelEDIFParser(compressed)) {
            Assertions.assertNotNull(p.parseEDIFNetlist());
        }
```

File: src/com/xilinx/rapidwright/edif/EDIFTools.java (hunk 1 above)
Comment (clavin-xlnx) [reply src/com/xilinx/rapidwright/edif/EDIFTools.java:758]: Not sure that this works in the uncompressed case.

Comment (eddieh-xlnx) [reply src/com/xilinx/rapidwright/edif/EDIFTools.java:758]: Oh yes, sorry! What about gating the inflation using the `.gz` check? My overall suggestion was that we can re-inflate the size back to the uncompressed value before calling the EDIF parser.

Comment (clavin-xlnx) [reply src/com/xilinx/rapidwright/edif/EDIFTools.java:758]: I don't think I understand the motivation of why this code needs to be modified.  The only thing this check does it decide between parallel parsing and serial parsing of the file based on size.

Comment (eddieh-xlnx) [reply src/com/xilinx/rapidwright/edif/EDIFTools.java:758]: To me, it's a little easier to understand/general if we always provide the uncompressed size to a method than to have something specific happen inside the parser. Scenarios like what if the exact size of the file becomes known, or if we have to support some other compression standard, becomes less intrusive to support. Only a suggestion, not anything to hold up this PR. Let's move on.

**Pull request merged**